landoapi/api/treestatus.py

```diff
@@ -0,0 +1,609 @@
+    """
+    session = db.session
+    tree = Tree.query.filter_by(tree=tree_name).one_or_none()
+    if not tree:
+        raise ProblemException(
+            404,
+            f"No tree {tree_name} found.",
+            "The tree does not exist.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404",
+        )
+    session.delete(tree)
+
+    # Delete from logs and change stack, too.
+    Log.query.filter_by(tree=tree_name).delete()
+    StatusChangeTree.query.filter_by(tree=tree_name).delete()
+    session.commit()
+    cache.delete_memoized(get_tree_by_name, tree_name)
+
+
+def update_tree_log(
+    id: int, tags: Optional[list[str]] = None, reason: Optional[str] = None
+):
+    """Update the log with the given id with new `tags` and/or `reason`."""
+    if tags is None and reason is None:
+        return
+
+    log = Log.query.get(id)
+
+    if log is None:
+        raise ProblemException(
+            404,
+            f"No tree log for id {id} found."
+            f"The tree log does not exist for id {id}.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404",
+        )
+
+    if tags is not None:
+        log._tags = json.dumps(tags)
+    if reason is not None:
+        log.reason = reason
+
+
+def get_combined_trees(trees: Optional[list[Tree]] = None) -> list[CombinedTree]:
+    """Return a `CombinedTree` representation of trees.
+
+    If `trees` is set, return the `CombinedTree` for those trees, otherwise
+    return all known trees.
+    """
+    query = (
+        Tree.query.distinct(Tree.tree)
+        .add_columns(
+            Log._tags,
+            Log.status,
+            Log.reason,
+            Log.id,
+        )
+        .outerjoin(
+            Log,
+            Log.tree == Tree.tree,
+        )
+        .order_by(Tree.tree.desc(), Log.created_at.desc())
+    )
+
+    if trees:
+        query = query.filter(Tree.tree.in_(trees))
+
+    return [get_combined_tree(*result) for result in query.all()]
+
+
+def _now() -> datetime.datetime:
+    return datetime.datetime.utcnow().replace(tzinfo=pytz.UTC)
+
+
+def update_tree_status(
+    session: sa.orm.Session,
+    tree: Tree,
+    status: Optional[str] = None,
+    reason: Optional[str] = None,
+    tags: Optional[list[str]] = None,
+    message_of_the_day: Optional[str] = None,
+):
+    """Update the given tree's status.
+
+    Note that this does not commit the session.
+    """
+    if not tags:
+        tags = []
+
+    if status is not None:
+        tree.status = status
+
+    if reason is not None:
+        tree.reason = reason
+
+    if message_of_the_day is not None:
+        tree.message_of_the_day = message_of_the_day
+
+    # Create a log entry if the reason or status has changed.
+    if status or reason:
+        if status is None:
+            status = "no change"
+        if reason is None:
+            reason = "no change"
+        log = Log(
+            tree=tree.tree,
+            created_at=_now(),
```

Comment:
This should be automatically filled by the `Base` model.

Reply:
This tiny change took a ton of effort to fix, since it's very hard to mock out timestamps when created by `db.func.now()`, as `db.func` is a dynamic DB-specific object in SQLAlchemy. I ended up re-writing most of the tests to use `pydantic` schema validation instead of just checking `dict == dict`, but in the end I think they're easier to understand and make changes to. :)